Content.Shared/Chemistry/Components/Solution.cs

```diff
@@ -674,58 +675,97 @@ public Solution SplitSolution(FixedPoint2 toTake)
                 {
-                    splitSolution.AddReagent(currentReagent);
-                    reagentsToRemove.Add(currentReagent);
+                    Contents[i] = new ReagentQuantity(reagent, quantity - toTakePer);
+                    newSolution.Contents.Add(new ReagentQuantity(reagent, toTakePer));
+                    Volume -= toTakePer;
                 }
                 else
                 {
-                    splitSolution.AddReagent(currentReagent.Reagent, toTakePer);
-                    RemoveReagent(currentReagent.Reagent, toTakePer);
+                    Contents.RemoveSwap(i);
+                    newSolution.Contents.Add(new ReagentQuantity(reagent, quantity));
+                    Volume -= quantity;
                 }
             }
 
-            foreach (var reagent in reagentsToRemove)
+            newSolution.Volume = origVol - Volume;
+            _heatCapacityDirty = true;
+            newSolution._heatCapacityDirty = true;
+
+            ValidateSolution();
+            newSolution.ValidateSolution();
+
+            return newSolution;
+        }
+
+        /// <summary>
+        /// Splits a solution, taking the specified amount of each reagent specified in reagents from the solution.
+        /// If any reagent in the solution has less volume than specified, it will all be transferred into the new solution.
+        /// </summary>
+        /// <param name="toTakePer">How much of each reagent to take.</param>
+        /// <returns>A new solution containing the reagents taken from the original solution.</returns>
+        public Solution SplitSolutionPerReagentWithOnly(FixedPoint2 toTakePer, params string[] reagents)
+        {
+            if (toTakePer <= FixedPoint2.Zero)
+                return new Solution();
+
+            var origVol = Volume;
+            Solution newSolution = new Solution(Contents.Count) { Temperature = Temperature };
+
+            for (var i = Contents.Count - 1; i >= 0; i--) // iterate backwards because of remove swap.
             {
-                RemoveReagent(reagent);
+                var (reagent, quantity) = Contents[i];
+
+                // Each reagent to split must be in the set given.
+                if (!reagents.Contains(reagent.Prototype))
+                    continue;
+
+                // If the reagent has more than enough volume to remove, no need to remove it from the list.
+                if (quantity > toTakePer)
+                {
+                    Contents[i] = new ReagentQuantity(reagent, quantity - toTakePer);
+                    newSolution.Contents.Add(new ReagentQuantity(reagent, toTakePer));
+                    Volume -= toTakePer;
+                }
+                else
+                {
+                    Contents.RemoveSwap(i);
+                    newSolution.Contents.Add(new ReagentQuantity(reagent, quantity));
+                    Volume -= quantity;
+                }
             }
-            if (Volume == FixedPoint2.Zero)
```

Comment:
we should still make this check and call

Reply:
Fine, but this does what?
```
      Contents.Clear();
      Volume = FixedPoint2.Zero;
      _heatCapacityDirty = false;
      _heatCapacity = 0;
```

Contents: If anything has a negative quantity, this may not be empty, sure.
Volume: Less than or equal to zero to start, fine, should never have negative values.
heatCapacityDirty: Overwritten, don't care.
_heatCapacity: Probably updated later with _heatCapacityDirty, likely don't care.